BUGFIX: show workspace owner/title in media usage tab

habuchholz opened the pull request:
Issue: https://github.com/neos/neos-development-collection/issues/5181

Possibility to enable showing all workspace titles/owners to user with certain roles in the media usage tab.

**Checklist**

- [X] Code follows the PSR-2 coding style
- [ ] Tests have been created, run and adjusted as needed
- [ ] The PR is created against the [lowest maintained branch](https://www.neos.io/features/release-roadmap.html)
- [X] Reviewer - PR Title is brief but complete and starts with `FEATURE|TASK|BUGFIX`
- [X] Reviewer - The first section explains the change briefly for change-logs
- [X] Reviewer - Breaking Changes are marked with `!!!` and have upgrade-instructions

habuchholz committed:
FIX: show workspace owner/title in media usage tab

- configurable privilege roles

Issue: https://github.com/neos/neos-development-collection/issues/5181

Sebobo commented on Neos.Media.Browser/Configuration/Settings.yaml:
```diff
@@ -27,6 +27,10 @@ Neos:
+        # By default, disable showing the workspace owner/title in media browser usage tab
+        showWorkspaceOwnerOrName:
+          enable: false
+          privilegedRoles: ['Neos.Neos:Administrator']
```
Any reason why you didn't introduce an actual privilege instead of defining a privilege via the settings?

You could introduce a new privilege and integrators can add it to whichever role they want.
Could also be enabled by default for admins IMO.

This way you only need to check if the new privilege is granted and simplify the code a bit.

habuchholz replied:
Good point, haven't thought about that one. 

I've implemented the privilege but im stuck trying to check the role without looping over them.
Would be nice to have something like `Account::hasRole()` including parent roles.

Or should i just loop over the user (parent) roles and match them with the new privilege role.

Sebobo replied:
You can either use `$this->securityContext->hasRole($roleIdentifier)` or `$this->privilegeManager->isPrivilegeTargetGranted($privilegeTarget);`

habuchholz replied:
Thanks worked like a charm!

habuchholz committed:
BUGFIX: introduce a new role for displaying the workspace owner

- new role `Neos.Media.Browser:WorkspaceName`
- add new role to `Neos.Neos:Administrator` as parent role
- enable showing the workspace owner by default
- adjust UsageController.php to check if current user role matches the new role

Issue: https://github.com/neos/neos-development-collection/issues/5181

habuchholz committed:
BUGFIX: remove unused variable in controller

- remove unused variable `$policyService` and import in UsageController.php and fix formatting
- correct commentary in Settings.yaml

Issue: https://github.com/neos/neos-development-collection/issues/5181

Sebobo commented on Neos.Neos/Configuration/Policy.yaml:
```diff
@@ -325,8 +325,7 @@ roles:
     description: Grants access to all modules and functionalities of the Neos backend.
-
-    parentRoles: ['Neos.Neos:Editor']
+    parentRoles: ['Neos.Neos:Editor', 'Neos.Media.Browser:WorkspaceName']
```
It would be better to assign the privilege in the Media.Browser package to not add a dependency here.
This is already done there for several other privileges like this example:
```yaml
  'Neos.Neos:Administrator':
    privileges:
      -
        privilegeTarget: 'Neos.Media.Browser:ManageAssetCollections'
        permission: GRANT
```

Sebobo commented on Neos.Media.Browser/Configuration/Settings.yaml:
```diff
@@ -27,6 +27,9 @@ Neos:
         createAssetRedirectsOption:
           enable: true
+        # By default, enable showing the workspace owner/title in media browser usage tab
+        showWorkspaceOwnerOrName:
```
IMO we don't need this switch. If someone doesn't want the feature they can disable or remove the privilege.

Sebobo commented on Neos.Media.Browser/Configuration/Policy.yaml:
```diff
@@ -29,6 +29,9 @@ privilegeTargets:
       matcher: 'method(Neos\Media\Browser\Controller\(Asset|Image)Controller->(replaceAssetResource|updateAssetResource)Action())'
 
 roles:
+  'Neos.Media.Browser:WorkspaceName':
```
If we just assign the privilege to the admin, we don't need this role IMO. If someone needs it, they can create their own.

habuchholz committed:
BUGFIX: introduce new privilege target `Neos.Media.Browser:WorkspaceName` instead of the role

- introduce new privilege target `Neos.Media.Browser:WorkspaceName`, grant privilege to `Neos.Neos:Administrator` and remove obsolet abstract role `Neos.Media.Browser:WorkspaceName`
- remove unnecessary parent role from `Neos.Neos:Administrator`
- remove config in Settings.yaml
- adjust UsageController.php to work with new privilege target

Issue: https://github.com/neos/neos-development-collection/issues/5181

Sebobo commented on Neos.Media.Browser/Configuration/Policy.yaml:
```diff
@@ -28,6 +28,10 @@ privilegeTargets:
 
+    'Neos.Media.Browser:WorkspaceName':
+      label: Allowed to see workspace owner
+      matcher: 'method(Neos\Media\Browser\Controller\UsageController->relatedNodesAction())'
```
But I think you now blocked access to the relatedNodes for everyone except admins and the check inside the method doesn't do anything as it's the same what the privilege does.

habuchholz replied:
We tried it with a normal Editor and he can access the media usage tab, but doesn't see the workspace titles (as expected). Do you maybe have another solution for the privilege?

skurfuerst replied:
maybe let's add a comment:

```
we only evaluate the WorkspaceName privilegeTarget in code (UsageController) to determine whether we should show workspace names or not. To control access to the module in general, Neos.Media.Browser:AssetUsage is used (so that's why they have the same matcher string).
```

Sebobo reviewed (changes requested):
We are getting there, thx for your work :)

I have an idea: you don't use the `isPrivilegedRole` variable but instead add a field `label` to the `inaccessibleRelation` entries.

There you use a new method `getLabelForInaccessibleWorkspace` which is checked by your `WorkspaceName` privilege. 
If `$isPrivilegedRole` is not given, then you leave the `label` for the relation empty for now (maybe we can make this nicer in the future and also add the translated part in there.
This way you can just render the label in the template the same way in all 3 cases.

Wdyt? Hope I don't discourage you with my feedback, but things like this are sometimes a bit hard to get right, even though they look simple.

habuchholz committed:
BUGFIX: new method to return label containing the workspace title/owner

- UsageController.php: introduce new method to return `label` for workspace title based on privilege target and workspace type
- RelatedNodes.html: use new `label` for rendering

Issue: https://github.com/neos/neos-development-collection/issues/5181

Jan3k3y commented:
Any updates here?🤔

Sebobo commented:
Everyone was on holidays 😉

Will take a look again and find some more reviews.

skurfuerst committed:
add minor comment when reviewing

skurfuerst reviewed (approved):
thanks everybody :) (I reviewed it once again together with @Jan3k3y synchronously)
